scalyr_agent/monitor_utils/k8s.py

```diff
@@ -566,42 +580,209 @@ def process_object( self, obj ):
+        self._cache_purge_secs=cache_purge_secs
+        self._config_update_count = 0
+
+        empty_config = _CacheConfig( api_url='', namespaces_to_ignore=[] )
 
-        self._k8s = k8s
 
-        self._namespace_filter = self._build_namespace_filter( namespaces_to_ignore )
-        self._node_filter = self._build_node_filter( self._namespace_filter )
+        new_config = _CacheConfig( api_url=api_url,
+                                   verify_api_queries=verify_api_queries,
+                                   cache_expiry_secs=cache_expiry_secs,
+                                   cache_purge_secs=cache_purge_secs,
+                                   namespaces_to_ignore=namespaces_to_ignore )
+
+        # perform initial configuration of the cache
+        self._configure( old_config=empty_config, new_config=new_config )
 
         # create the controller cache
-        self._controller_processor = ControllerProcessor( logger )
-        self._controllers = _K8sCache( logger, k8s, self._controller_processor, '<controller>',
-                               filter=self._namespace_filter,
+        self._controller_processor = ControllerProcessor()
+        self._controllers = _K8sCache( self._controller_processor, '<controller>',
                                perform_full_updates=False )
 
         # create the pod cache
-        self._pod_processor = PodProcessor( logger, self._controllers )
-        self._pods = _K8sCache( logger, k8s, self._pod_processor, 'Pod',
-                               filter=self._node_filter )
+        self._pod_processor = PodProcessor( self._controllers )
+        self._pods = _K8sCache( self._pod_processor, 'Pod' )
 
         self._cluster_name = None
-        self._cache_expiry_secs = cache_expiry_secs
-        self._cache_purge_secs = cache_purge_secs
         self._last_full_update = time.time() - cache_expiry_secs - 1
 
-        self._lock = threading.Lock()
         self._initialized = False
 
-        self._thread = StoppableThread( target=self.update_cache, name="K8S Cache" )
+        self._thread = None
+
+        if start_caching:
+            self.start()
+
+    def _configure( self, old_config, new_config ):
+        """
+        Configures the cache based on any changes in the configuration
+        @param old_config: the old configuration
+        @param new_config: the new configuration
+        @type old_config: _CacheConfig
+        @type new_config: _CacheConfig
+        @return: True if the config changed, False otherwise
+        """
+
+        changed = False
+
+        # get old state values
+        old_k8s = None
+        old_namespace_filter = None
+        old_node_filter = None
+        old_config_update_count = 0
+        self._lock.acquire()
+        try:
+            old_k8s = self._k8s
+            old_namespace_filter = self._namespace_filter
+            old_node_filter = self._node_filter
+            old_config_update_count = self._config_update_count
```

Comment:
I'm pretty sure there is a flaw in using the `config_update_count` approach to detecting if some other thread has updated the cache.  Let's say this happens, two threads are trying to update the cache config at the same time.  Which thread should win?  I would argue that it's the thread that invoked this method last (last caller wins).  Your approach doesn't guarantee that.

It's a little hard to show, but let me try.  Suppose Thread A invokes `_configure` first.  It will get to this critical block first and record `old_config_update_count` as 5.  It then releases the lock to do its work to build up the filters.  Thread B now invokes `_configure`.  It has invoked it after A.  It also gets to the first critical block and records `old_config_update_count` as 5.  It then releases the lock to build the filters.... Well, suppose A then actually updates the cache and increments `_config_update_count` to 6.  B's update to the cache config will fail.. breaking the invariant that I want the latest call to win.

Reply:
Done.